rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -191,42 +191,52 @@ private class InMemHotStore[F[_]: Sync, C, P, A, K](
   def putJoin(channel: C, join: Seq[C]): F[Unit] =
     for {
       joins <- getJoins(channel)
-      _ <- if (!joins.contains(join)) S.flatModify { cache =>
-            Sync[F].delay(cache.joins.put(channel, join +: joins)).as(cache)
+      _ <- if (!joins.contains(join)) S.modify { cache =>
+            ignore(cache.joins.put(channel, join +: joins))
+            (cache, ())
           } else Applicative[F].unit
     } yield ()
 
-  def installJoin(channel: C, join: Seq[C]): F[Unit] = S.flatModify { cache =>
-    Sync[F]
-      .delay {
-        val installed = cache.installedJoins.get(channel).getOrElse(Seq.empty)
+  def installJoin(channel: C, join: Seq[C]): F[Unit] = S.modify { cache =>
+    ignore {
+      val installed = cache.installedJoins.get(channel).getOrElse(Seq.empty)
 
-        if (!installed.contains(join))
-          cache.installedJoins
-            .put(channel, join +: installed)
-      }
-      .map(_ => cache)
+      if (!installed.contains(join))
+        cache.installedJoins
+          .put(channel, join +: installed)
+    }
+    (cache, ())
   }
 
   def removeJoin(channel: C, join: Seq[C]): F[Unit] =
     for {
       joins         <- internalGetJoins(channel)
       continuations <- getContinuations(join)
-      _ <- if (continuations.isEmpty) S.flatModify { cache =>
-            val index = cache.joins(channel).indexOf(join)
-            if (index != -1) removeIndex(cache.joins(channel), index) >>= { updated =>
-              Sync[F]
-                .delay {
-                  cache.joins.put(channel, updated)
-                }
-                .as(cache)
-            } else cache.pure[F]
-          } else Applicative[F].unit
+      s             <- S.get
+      index         = s.joins(channel).indexOf(join)
+      _ <- if (continuations.isEmpty && index != -1)
+            checkIndex(s.joins(channel), index) >>
+              S.modify { cache =>
+                val updated =
+                  removeIndex(
+                    cache.joins(
+                      channel
+                    ),
+                    index
+                  )
+                ignore(
+                  cache.joins.put(
+                    channel,
+                    updated
+                  )
+                )
+                (cache, ())
+              } else Applicative[F].unit
```

Comment:
I use the `Applicative[F].whenA` or `unlessA` pattern - I find it more readable and terser. But YMMV